agent/api/task/task.go

```diff
@@ -2382,15 +2395,41 @@ func (task *Task) dockerPortMap(container *apicontainer.Container, dynamicHostPo
 		}
 	}
 
+	// For each port binding config, either one of containerPort or containerPortRange is set.
+	// (1) containerPort is the port number on the container that's bound to the user-specified host port or the host port assigned by ECS Agent.
+	// (2) containerPortRange is the port number range on the container that's bound to the mapped host port range found by ECS Agent.
+	var err error
 	for _, portBinding := range containerToCheck.Ports {
-		// for each port binding config, either one of containerPort or containerPortRange is set
 		if portBinding.ContainerPort != 0 {
 			containerPort := int(portBinding.ContainerPort)
+			protocolStr := portBinding.Protocol.String()
+			dockerPort := nat.Port(strconv.Itoa(containerPort) + "/" + protocolStr)
 
-			dockerPort := nat.Port(strconv.Itoa(containerPort) + "/" + portBinding.Protocol.String())
-			dockerPortMap[dockerPort] = append(dockerPortMap[dockerPort], nat.PortBinding{HostPort: strconv.Itoa(int(portBinding.HostPort))})
+			if portBinding.HostPort != 0 {
+				// An User-specified host port exists
+				hostPortStr = strconv.Itoa(int(portBinding.HostPort))
```

Comment:
where do we validate that this HostPort is a valid int?

Reply:
ECS front end service will validate it when customers create a task definition. For ECS front-end service, an valid host port is
1. `nil`, which means that customers do not specify host port for port mapping in their task definitions.
2. `0`,  which also means that customers do not specify host port for port mapping in their task definitions.
3. `value > 0` and `value <= MAX_PORT_NUMBER`,  and the MAX_PORT_NUMBER is set to `65535`.

As Agent trusts payload messages streamed down from ECS, and this error should be caught before streaming down to Agent,  we don't perform `portBinding.HostPort` validation here as well.

Reply:
#### Test cases

- Case 1. Attempt to register a task definition with `hostPort: -1` using ECS console
- Case 2. Attempt to register a task definition with `hostPort: -2147483648` using ECS console
- Case 3. Attempt to register a task definition with `hostPort: 2147483648` using ECS console
- Case 4. Attempt to register a task definition with `hostPort: 0.1` using ECS console
- Case 5. Attempt to register a task definition with `hostPort: "111"` using ECS console

#### Results

- Case 1, 2, and 3. Failed to register the task definition. Receive the error message "Invalid 'hostPort' setting for container 'xxx'".
- Case 4. Successfully registered the task definition, and the hostPort is round to the 0.
- Case 5. Successfully registered the task definition, and the hostPort is 111 (`""` is removed by ECS front end).